colosseum/fonts.py

```diff
@@ -0,0 +1,93 @@
+import re
+
+from .constants import (FONT_STYLE_CHOICES, FONT_VARIANT_CHOICES,
+                        FONT_WEIGHT_CHOICES, INHERIT, INITIAL_FONT_VALUES,
+                        NORMAL, SYSTEM_FONT_KEYWORDS)
+from .validators import ValidationError
+
+# Find spaces within single and double quotes
+SQ_PATTERN = re.compile(r"\s+(?=(?:(?:[^']*'){2})*[^']*'[^']*$)")
+DQ_PATTERN = re.compile(r'\s+(?=(?:(?:[^"]*"){2})*[^"]*"[^"]*$)')
+
+
+def get_system_font(keyword):
+    """Return a font object from given system font keyword."""
+    if keyword in SYSTEM_FONT_KEYWORDS:
+        # Get the system font
+        font = INITIAL_FONT_VALUES.copy()
+    return font
+
+
+def construct_font_property(font):
+    """Construct font property string from a dictionary of font properties."""
+    return ('{font_style} {font_variant} {font_weight} '
+            '{font_size}/{line_height} {font_family}').format(**font)
+
+
+def replace_font_family_spaces(string, space_sep):
+    """Replace spaces between quotes by character."""
+    string = SQ_PATTERN.sub(space_sep, string)
+    string = DQ_PATTERN.sub(space_sep, string)
+    return string
+
+
+def parse_font_property(string):
+    """
+    Parse font string into a dictionary of font properties.
+
+    Reference:
+    - https://www.w3.org/TR/CSS22/fonts.html#font-shorthand
+    - https://developer.mozilla.org/en-US/docs/Web/CSS/font
+    """
+    font = INITIAL_FONT_VALUES.copy()
+    comma, space = ', ', ' '
+    comma_sep, space_sep = '~', '@'
+
+    # Remove extra inner spaces
+    string = space.join(string.strip().split())
+
+    # Replace commas between font families with some character
+    string = string.replace(comma, comma_sep)
+    string = string.replace(comma[0], comma_sep)
+    string = replace_font_family_spaces(string, space_sep)
```

Comment:
Since all these are options, and there could be several font families and some font families have quotes (if using weird characters or using spaces) I thought the easiest would be to replace the comma and space there and the spaces within quotes and then do a simple split.

```
font: 1.2em "Fira Sans", sans-serif;
font: italic 1.2em "Fira Sans", serif;
font: italic small-caps bold 16px/2 cursive;
font: small-caps bold 24px/1 sans-serif;
font: caption;
```

>Since we're already knee deep with a complex regex

I would say it does not help much since the regex would end up being uber complex (AFAICT)

Reply:
I am open for suggestions here :-)

Reply:
Ok - but I think we can work the underlying structure of the font specifier to our advantage here.

The font family must be the the last "element" in the specifier; everything before that point is either from a "known keyword" list (`bold`/`italic`/...) or fits a known pattern (`\dpx`,...). My approach would be to `lsplit(' ', 1)` from the front of the specifier trying to match the "known" patterns until you find a split that won't work - at which point, you take the version of the string prior to the last `lsplit`, and try and parse it as a font family.

Reply:
That could work as well :-)

Reply:
Ok, added the suggested fix